### DISCUSSION TITLE
Labs Feedback: @lit-labs/preact-signals

### DISCUSSION THREAD

**Discussion by justinfagnani** [lit/lit#4115] [Labs Feedback]
The `@lit-labs/preact-signals` package is a set utilities for using [Preact Signals](https://github.com/preactjs/signals) in Lit.

- `SignalWatcher`: a class mixin that makes an element watch all signal accesses during the element's reactive update lifecycle, and triggers an element update when signals change.
- `watch()`: a lit-html directive that accepts a single Signal and renders its value, subscribing to updates and updating the DOM when the signal changes.
- `html`, `svg`, and `withWatch()`: `html` and `svg` template tags that can be used in place of Lit's default template tags which automatically wraps any signals in `watch()`. `withWatch()` applies the auto-watching behavior to any lit-html compatible template tag.

We want to collect your feedback on your experiences with `@lit-labs/preact-signals` as we iterate on it according to our Lit Labs process and move the package towards possible graduation or retirement.

- Status: **Prototyping**
- Major known issues: None yet
- Issues: [[labs/preact-signals]](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+%5Blabs%2Fpreact-signals%5D)
- Docs: [README](https://github.com/lit/lit/tree/main/packages/labs/preact-signals#readme)
- npm: [@lit-labs/preact-signals](https://www.npmjs.com/package/@lit-labs/preact-signals)

Please try `@lit-labs/preact-signals` and tell us about your experience. Leave comments with your thoughts; highlight what issues or missing features you're blocked on; give a link to a project that uses it; etc.

And so we can have some numbers to look at, you can add reactions to this comment to indicate your experience with this package:

🚀 I have tried this package and it works great. I need it. Ship it!
👍 I have tried this package and it seems like a good idea.
😕 I have tried this package and would use it, but it has serious bugs or missing features.
👀 I am interested, but haven't tried the package yet.
👎 I'm not interested in this package.

Note: other reactions like 😄, 🎉, and ❤️ don't mean anything as part of this scheme. You probably mean, and please use, 🚀, 👍, or  👀.

**Comment by andyjessop** [comment]
I'm getting a 404 on the docs link.

**Reply by marcomuser** [reply]
It's part of the 3.0 branch: https://github.com/lit/lit/tree/3.0/packages/labs/preact-signals

**Reply by AndrewJakubowicz** [reply]
`@lit-labs/preact-signals` has not been published yet. Will update the discussion once a pre-release has been published.

**Reply by AndrewJakubowicz** [reply]
It has been published in a pre-release! [Blog post announcement](https://lit.dev/blog/2023-09-27-lit-3.0-prerelease-2/)!

You can now try the package with: `npm i @lit-labs/preact-signals`

**Comment by danielquintero** [comment]
Thanks for putting this together, looks promising. I just have one question, are there any plans to support deepSignals too? For reference (https://github.com/preactjs/signals/issues/4)

**Reply by redfox-mx** [reply]
Hi @danielquintero watch directive and Signals Watcher uses preact tracking system (subscription and effects), in that sense any new feature with those approaches must be supported. Don't worry about.

**Comment by jogibear9988** [comment]
Is the source anywhere available?

**Reply by redfox-mx** [reply]
Yes! You can see source code here https://github.com/lit/lit/tree/main/packages/labs/preact-signals

PD: Hi @jogibear9988 n.n

**Comment by herrKlein** [comment]
Yes, really cool, we already use it in production! we did a combination of litlabs context, where the context is the signal

**Reply by marcomuser** [reply]
What's the benefit of combining signals with a context? You can already freely import the signal wherever you need to.

**Reply by redfox-mx** [reply]
Ooooh sounds really interesting. So, when you call a context, are you subscribing to an effect? Or how did you design signal/context pattern?

Maybe you pass a signal into context subscription or emit a value on response to a signal change?

**Reply by herrKlein** [reply]
haha, good question.. made me think. Maybe its not necessary, but we started out with a state object and using litcontext, children could consumer this context. Later on we tried signals, and copied the signal value to the context. After that we thought it was simpler to just made the signal be the context. Just learning signals, what is a logic place for the signal to live? outside a webcomponent?

**Reply by marcomuser** [reply]
I would put signals outside of components into the module scope. From there you can import them wherever you want to and then subscribe to their value from within components. But I guess there are many ways to rome. 😀

**Reply by klauss194** [reply]
@marcomuser can you share a bit of how did you do that ? we're struggling to share signals from one module to another ( we have vaadin router mounting routes as well )

**Comment by louwers** [comment]
When passing a signal to a component you need to use

```
<my-component .someProp=${mySignal}
```

instead of

```
<my-component .someProp=${mySignal.value}
```

this should be clearer in the documentation.

**Reply by louwers** [reply]
This is pretty annoying because it breaks type checking.

**Reply by rstoneIDBS** [reply]
I'm new to signals, just looking into this package but I wonder why would you pass the signal down into the component rather than referencing it directly from within the component? Isn't this how signals are meant to be consumed?

**Reply by louwers** [reply]
I don't want to pass the signal. I want to pass the value of the signal to a reusable-component that is unaware of signals.

**Reply by justinfagnani** [reply]
There are actually many ways to pass a signal to a child component, depending on which utility from `@lit-labs/preact-signals` you're using, or whether the child component directly accepts signals:

- `.someProp=${mySignal.value}` if you're using the `SignalWatcher` mixin
- `.someProp=${watch(mySignal)}` if you're using `watch()` directive
- `.someProp=${mySignal}` if you're using the `html` tag from `@lit-labs/preact-signals` 
- `.someProp=${mySignal}` if the child component directly accepts signals

We have all of these options because we're trying to gather feedback on which are best suited for people's use cases. It may turn out that all of them are good for different situations or preferences.

`.someProp=${mySignal}` does indeed not yet type-check with the lit-analyzer. We would have to figure out how to make that work. Maybe there is some type that signal `html` tag could implement to indicate that it can take a `Signal<T>` where a `<T>` is expected. I haven't looked into this yet.

**Reply by louwers** [reply]
Thanks for your reply!

**Reply by louwers** [reply]
The type checking problem is not just a problem for lit-analyzer. When using a simple function that returns a template, it's also broken.

```ts
import { signal, html } from "@lit-labs/preact-signals";

export function wrapper() {
  const text = signal("My header");
  setTimeout(() => {
    text.value = "Timeout reached";
  }, 3000);

  return html`
    ${header(text)}
          // ^ ERROR!
  `;
}

function header(header: string) {
  return html`
    <h1>${header}</h1>
  `;
}
```

It would be much cleaner if `html` from `@lit-labs/preact-signals` would work with `text.value` in the above example, or if `Signal<T>` was somehow made assignable to `T`.

**Reply by justinfagnani** [reply]
That error is correct. You can't pass a signal to a function expecting a string. You need to pass `text.value`. This has nothing to do with Lit, that `header()` function call is plain JavaScript.

**Reply by louwers** [reply]
@justinfagnani I agree that it should be possible to pass a signal’s value to a plain JavaScript function. But the only way to make this reactive is by passing the signal. Passing `text.value` makes the example non-functional.

**Reply by justinfagnani** [reply]
Then you need to make the function take a signal and not a string. That `header(text)` call doesn't involve Lit at all, it's just JS. There is nothing Lit can do to make that work.

**Reply by tbroyer** [reply]
Or use a `computed()`?

```ts
import { signal, computed, html } from "@lit-labs/preact-signals";

export function wrapper() {
  const text = signal("My header");
  setTimeout(() => {
    text.value = "Timeout reached";
  }, 3000);

  const h = computed(() => header(text.value));

  return html`
    ${h}
  `;
}

function header(header: string) {
  return html`
    <h1>${header}</h1>
  `;
}
```

**Reply by louwers** [reply]
It would be great if all template literal parameters that are functions were automatically wrapped in `computed`.

This is essentially what ArrowJS does (framework inspired by Lit and Vue).

The above would simplify to:

```js
html`${() => header(text.value)}`
```

**Reply by louwers** [reply]
Thanks for your suggestion @tbroyer.

I would still like to see some kind of automatic `computed` wrapping. Maybe it is even possible to do this yourself with a custom directive? You'd need to differentiate between functions that are passed to a click handler and those that are passed for the template they return.

In any case, I love writing 'templates' like this over the class-based syntax.

```js
export function exercisesOverviewPage() {
  const exercises = getDocsSignal("exercise");

  return html`
    ${computed(
      () => html`
        <div class="flex flex-col items-center text-center">
          ${exercises.value.map(
            ({ name, _id }) => html`
              <a
                class="bg-light w-sm m-2 p-3 rounded-8 select-none"
                @click=${() => goToURL(`/exercise/${_id}`)}
              >
                ${name}
              </a>
            `
          )}
        </div>
      `
    )}
  `;
}
```

**Reply by tidemann** [reply]
Would be nice to see an actual example that demonstrates how to use withWatch. I looked in README, but withWatch doesn't seem to be used in the provided example.

Update: My bad. Just noticed html was imported from preact-signals, and not from lit in that example.

**Comment by krazylek** [comment]
Hi, when using the `watch` directive, I find quite difficult to work with signals that are Array or Object.

Simple operations like the following example aren't obvious:

```ts
const list = signal(['a', 'b']);

return html`<p>The list values are ${watch(list).value.join(', ')}</p>`; // doesn't work as `watch()` returns a directive
```
Especially true when compared to the other `SignalWatcher` or signals' `html` methods:

```ts
return html`<p>The list values are ${list.value.join(', ')}</p>`; // very straightforward!
```

Reading more, I discovered that using `computed` would make it work as expected:

```ts
const listToString = computed(() => list.value.join(', '));

return html`<p>The list values are ${watch(listToString)}</p>`; // does work!
```

But do I need to create a computed each time I want to use a nested value in a `watch`?
Did I miss another method?

**Reply by sorvell** [reply]
Here's a [todo example](https://lit.dev/playground/#gist=b2069a4710633eda1e4a3ecab13ae8ee) that uses signals with arrays.

You're not explicitly hitting this here, but we always need to start with: `Objects` and `Arrays` are tricky because mutating them (e.g. `list.push(...)`) doesn't change their identify and libraries like Lit and signals detect changes via identify.

You have a couple of options with watch/SignalWather, depending on your needs. 
* Using `watch` ensures you get a super granular sub-component update that can maximize performance. Typically this isn't needed since Lit by default is very fast at updates. And as you note, you must give `watch` a signal and return its output to the part and this means you might need to create a computed signal... fwiw re-creating the computed signal is probably fine in this case)
* However, just to render correctly, accessing the value in the render method while using `SignalWatcher` is all you need (i.e. just remove the use of `watch`).

Now, again, this only works if you actually change the referenced value (e.g. `list.value.push(...)` _bad_, `list.value = [...list.value, newItem]` _good_). 

Hopefully that's clear, but happy to give more info if needed.

**Reply by krazylek** [reply]
Thank you for your answer. Yes a very granular update was what I was looking.

The other options are nice as well, but I kept exploring a way to achieve this "granular transformation"

First I noticed that a `computed` does not need to be attached to a property or variable.
This make this alright:

```ts
const list = signal(['a', 'b']);
html`<p>The list values are ${watch(computed(() => list.value.join(', ')))}</p>`; 
```

Looks very close to what I was looking! Just wondering is that could lead to some leak, for example if a `LitElement` having such syntax is removed from the DOM, will that clear the computed signal?

Also is adding some sugar to `watch`, to allow a compute function as parameter, something that could be considered?
Eg.

```ts
html`<p>The list values are ${watch(() => list.value.join(', '))}</p>`; 
```

I know I could create my own `watch` function to do that, but it would be nice to have a standard way of doing it!

**Comment by e111077** [comment]
## Problem

Working with control flow with the `watch()` directive needs a helper. Playing around with Preact Signals + Lit I found something similar to what other people have been mentioning about dealing with arrays. e.g. consider the following case:

```ts
const list = signal([{val: 'a', id: 0}, {val: 'b', id: 1}, {val: 'c', id: 2}]);

export class MyElement extends LitElement {
  render() {
    return html`
      <div>
        Things in the list: ${
          watch(computed(() => repeat(list.value, item => item.id, (item, index) => html`
            <div>${item.val}</div>
            <!-- this is inside a computed function, so this does not need another `watch()` to trigger -->
           ${index !== item.length - 1 ? html`<div class="divider"></div>` : nothing}
          `))
        }
      </div>
      <div>
        <!-- this ternary needs a computed to trigger recalculation --> 
        The number of items is ${watch(computed(() => (list.value.length % 1 === 0 ? 'even' : 'odd')))}
      </div>
    `;
  }
}
```

`watch()` only accepts a `Signal` value. Therefore your expression needs to return a `Signal` value in order for the control flow to be recalculated whenever the value changes. In the case above, I wrapped everything in a `computed()` wrapper to return a single signal.

## Proposed Solution

I've got two approaches off the top of my head:

### Vend a `flow()` helper

Maybe just have a `flow()` (or whatever bikeshedded name) helper to handle things like control flow or things that use Signals but don't return a Signal which just does the following:

```ts
export function flow(val: () => unknown) {
  return watch(computed(val));
}
```

This would result in the following code:

```ts
const list = signal([{val: 'a', id: 0}, {val: 'b', id: 1}, {val: 'c', id: 2}]);

export class MyElement extends LitElement {
  render() {
    return html`
      <div>
        Things in the list: ${
          flow(() => repeat(list.value, item => item.id, (item, index) => html`
            <div>${item.val}</div>
            <!-- this is inside a computed function, so this does not need another `watch()` to trigger -->
           ${index !== item.length - 1 ? html`<div class="divider"></div>` : nothing}
          `)
        }
      </div>
      <div>
        <!-- this ternary needs a computed to trigger recalculation --> 
        The number of items is ${flow(() => (list.value.length % 1 === 0 ? 'even' : 'odd'))}
      </div>
    `;
  }
}
```

### Pros

- Simple implementation
- Named for the cases for which it is used
- `watch()` calls inside the `flow()` become unnecessary

### Cons

- You have to think about when to use `watch()` and when to use `flow()`
- You have to be aware that the entire `flow(fn)` fn value is going to be recalculated whenever a signal inside of it is changed which may lead to some unnecessary recalcs that could have been solved with a `watch()`
- `watch()` calls inside the `flow()` are superfluous so it's inconsistent (perhaps watch() should also support functions to prevent whole template recalc?)

### Enhance the `watch()` directive

Perhaps allow `watch()` to accept a `Signal|()=>unknown`, check if it's a signal. If so, subscribe to it. Otherwise, wrap it in a `computed()`

### Pros

- Same API
- Generally don't have to think about it

### Cons

- A bit leaky as you need to know when to wrap in a function or not

**Reply by e111077** [reply]
oh additionally, `html` signal template tag won't work for all use cases here in the case that you want to actually pass a signal value to another element. e.g.

```ts
import {html, signal, watch} from '@lit-labs/preact-signals';

const countSignal = signal(0);
const users = signal(['a', 'b', 'c']);

// using preact signals html tagged template
html`
  <div>
    Counters:
    ${
      users.value.map(user => html`
        <div>
          ${user}:
          <!-- does not pass signal reference -->
          <my-counter .count=${countSignal}></my-counter>
         </div>
      `)
    }
  </div>
`
```

In the case above, the `.count=${countSignal}` does not do what you expect and pass a signal reference. Instead the html template tag will attempt to pass the current value of the `countSignal` Meaning that all rendered `<my-counter>`s will not be synchronized since they instead have raw values passed to them.

**Reply by justinfagnani** [reply]
So I'm building another signals package right now to integrate with the [TC39 Signals proposal](https://github.com/proposal-signals/proposal-signals). One issue I'm hitting is that I can't implement a synchronous `watch()` directive (See https://github.com/proposal-signals/proposal-signals/issues/121).

If this situation holds, it means we'll have two choices for `watch()`:
1. It updates async, but uncoordinated with the element's update lifecycle, like `until()`. `el.updateComplete()` will not consider to `watch()` calls.
2. It calls `requestUpdate()` to be part of the element's batch, but this means it's no different than `const watch = (s) => s.get()` and using `SignalWatcher`.

I'm not sure if `watch()` is worth it at that point.

Targeted, but uncoordinated async writes are heard to deal with, and is why re now recommend Task over `until()`. We might not want to introduce a new `until()` that might even be more pervasively used. "How do I know when my element is done updating if I use `watch()`?" _will_ be a common question.

Calling `requestUpdate()` means that we don't really have updates that are any more specifically targeted than we already do, so `watch()` is kind of pointless.

So, we need to think that part through. It's still possible that even without `watch()` that we'd want a signal-friendly utility for looping though, since if we can detect what part of a list changed we can do a more targeted update, even with `SignalWatcher`. I'm not sure what that would look like yet, but maybe we should focus there so we have some benefit even if a synchronous `watch()` isn't feasible.

**Reply by e111077** [reply]
Doesn’t this make the html tag integration gross as well? You should report your findings on the proposal repo for their consideration

**Reply by justinfagnani** [reply]
Yeah, the html tag becomes just sugar so you don't have to write `signal.get()` everywhere. That alone might still be useful, at least if it's integrated into the template type-checker.

**Reply by sorvell** [reply]
I was experimenting with this as well and hit the same notify/effect sync issue. I just queued a microtask which worked but was in-elegant: https://lit.dev/playground/#gist=b2069a4710633eda1e4a3ecab13ae8ee

**Reply by sorvell** [reply]
@e111077 
I don't see how `flow(...)` is much better than `watch(computed(...))`. The boilerplate savings doesn't seem worth the obfuscation. Having `watch` auto-wrap in a computed seems better but agreed that you have to know when to pass a function.

@justinfagnani 
I think not supporting sub-component granular updates would be really disappointing. Most of the time it's unneeded, but (1) it can occasionally be useful, (2) I think it's a core expectation of signal users (everyone touts this as a selling point.. let's manage our novelty budget here), (3) while alternate good patterns are possible they may be in tension with (2).

**Reply by justinfagnani** [reply]
@sorvell the thing is that most signal-based frameworks do batching of signal updates, so in the end they're not much different from Lit. In a batch update, we iterate through parts to see which have changed, they iterate through the changed signals. It's computationally similar, _except_ for repeats where signals can short-cut evaluation of the iterable. That's where I think we could find some improvements.

**Reply by e111077** [reply]
@justinfagnani 
The difference that I see here is that they have the potential to replace the reactive element lifecycle with something in the platform just as we do with the custom element lifecycle to handle attributes. A `watch()` directive approach would give people an easier upgrade path.

@sorvell 
Please tell me this is not pain `watch(computed(() => repeat(`

**Reply by justinfagnani** [reply]
@e111077 Signals would not replace the reactive element lifecycle. They specifically leave scheduling up to "frameworks" and all signal-based frameworks now implement batching not to dissimilar from our lifecycle.

The intention is for a library to use primitive to watch signals, then schedule an update, then read the signals. The whole tension with the signals primitive for us is that they don't allow reading a signal during notification, so you _have_ to read them async. Their examples show doing this by enqueing a microtask, which is more or less what we do already.

See their example here: https://github.com/proposal-signals/proposal-signals?tab=readme-ov-file#implementing-effects

```ts
// This function would usually live in a library/framework, not application code
// NOTE: This scheduling logic is too basic to be useful. Do not copy/paste.
let pending = false;

let w = new Signal.subtle.Watcher(self => {
    if (!pending) {
        pending = true;
        queueMicrotask(() => {
            pending = false;
            for (let s of self.getPending()) s.get();
            self.watch();
        });
    }
});

// An effect effect Signal which evaluates to cb, which schedules a read of
// itself on the microtask queue whenever one of its dependencies might change
export function effect(cb) {
    let destructor;
    let c = new Signal.Computed(() => { destructor.?(); destructor = cb(); });
    w.watch(c);
    c.get();
    return () => { destructor.?(); w.unwatch(c) };
}
```

**Reply by e111077** [reply]
I mean perhaps we can have a future where the RE lifecycle heavily relies on signals in a way that reduces bytes and relies more on the platform for reactivity rather than our current approach.

There would definitely need to be some work to convert attributes to signals / update signals as well as integrate reactive controllers, and we’d definitely need to weigh the performance and size. But a standardization here could mean that this is what developers might eventually expect to interact with reactivity, so we should consider what a deeper integration would look like and send feedback to make sure we have that option open in the future.

**Reply by sorvell** [reply]
@justinfagnani Unless I'm mistaken, it doesn't seem computationally similar to me. If I have an element with N parts and 1 signal changes, if I use 'watch', it will avoid ~N dirty checks. If N is say, 100k, that could be significant. Is that wrong?

@e111077 Ack the DX issue, just trying to find the best solution. I wonder if maybe the auto-function thing + a signal/directive wrapper of some sort might be good... so you could just do `repeat` (signal version) and it'd work (somehow)?

**Reply by sorvell** [reply]
Explored some of these ideas in this [playground](https://lit.dev/playground/#gist=18772babe1862dd5437c04a41e57c9cc).

@justinfagnani Ack the issue with uncoordinated async updates. Perhaps this can be addressed. This would need more thought but here's a sketch. In the example above...
1.  signal effects are queued and there's a `signalEffectsComplete()` helper. 
2. updateComplete awaits this helper.
3. the element's own `updated` is pushed off by this helper as well (hacked this in via `_$didUpdate` atm).

We'd need to determine how to auto-integrate this better but then, (1) the element itself (e.g. updated) wouldn't have to do anything special to understand when it rendered, (2) anyone using updateComplete would also get the expected element update state.

@e111077 Also in the example above, the special signal `html` always wraps functions and directives in addition to signals. Functions that aren't mapped to event parts become computed. I didn't do anything yet with directives.

**Comment by zuhairtaha** [comment]
I wonder why computed doesn't work with i18next in this example:
```ts
import { computed, html } from "@lit-labs/preact-signals";
import i18next from "i18next";
import { LitElement } from "lit";
import { customElement } from "lit/decorators.js";

i18next.init({
  lng: "en",
  resources: {
    en: { translation: { hi: "Hi" } },
    es: { translation: { hi: "Hola" } },
  },
});

export function t(key: string) {
  return computed(() => i18next.t(key));
}

@customElement("my-element")
export class MyElement extends LitElement {
  render() {
    return html`
      <div>${t("hi")}</div>
      <button @click=${() => i18next.changeLanguage("es")}>es</button>
    `;
  }
}

```
I'm using `html` from @lit-labs/preact-signals
Also SignalWatcher and watch doesn't work

**Comment by sorvell** [comment]
Here's a working [example](https://lit.dev/playground/#gist=f6add86e6812682c4dd1a1452bc7737d).

The issue here is that signals track changes only in other signals. Your computed signal is not tracking any signals so it never re-computes. Since the i18next library provides a notification for when the language changes, this just needs to be exposed as a signal that's referenced in the computed. Here's the relevant code from the example:

```ts
export const t  = (key: string | (() => string)) => {
  return computed(() => {
    changes.value;
    return i18next.t(typeof key === 'function' ? key() : key);
  });
}

const changes = signal(0);
export const changeLanguage = (s: string) => {
  i18next.changeLanguage(s, () => {
    changes.value++;
  });
}
```
